common/src/Model/TagMatcher.cpp

```diff
@@ -98,13 +237,83 @@ namespace TrenchBroom {
             return std::make_unique<EntityClassNameTagMatcher>(m_pattern);
         }
 
-        bool EntityClassNameTagMatcher::matches(const Brush& brush) const {
-            const auto* entity = brush.entity();
-            if (entity == nullptr) {
-                return false;
+        bool EntityClassNameTagMatcher::matches(const Taggable& taggable) const {
+            BrushMatchVisitor visitor([this](const Brush& brush) {
+                const auto* entity = brush.entity();
+                if (entity == nullptr) {
+                    return false;
+                }
+
+                return matchesClassname(entity->classname());
+            });
+
+            taggable.accept(visitor);
+            return visitor.matches();
+        }
+
+        void EntityClassNameTagMatcher::enable(TagMatcherCallback& callback, MapFacade& facade) const {
+            if (!facade.selectedNodes().hasOnlyBrushes()) {
+                return;
             }
 
-            return StringUtils::caseInsensitiveMatchesPattern(entity->classname(), m_pattern);
+            const auto& definitionManager = facade.entityDefinitionManager();
+            const auto& allDefinitions = definitionManager.definitions();
+            auto matchingDefinitions = Assets::EntityDefinitionList{};
+
+            std::copy_if(std::begin(allDefinitions), std::end(allDefinitions), std::back_inserter(matchingDefinitions), [this](const auto* definition) {
+                return definition->type() == Assets::EntityDefinition::Type_BrushEntity && matchesClassname(definition->name());
+            });
+
+            std::sort(std::begin(matchingDefinitions), std::end(matchingDefinitions), [](const auto* lhs, const auto* rhs) {
+                return StringUtils::caseInsensitiveCompare(lhs->name(), rhs->name()) < 0;
+            });
+
+            const Assets::EntityDefinition* definition = nullptr;
+            if (matchingDefinitions.empty()) {
+                return;
+            } else if (matchingDefinitions.size() == 1) {
+                definition = matchingDefinitions.front();
+            } else {
+                StringList options;
+                std::transform(std::begin(matchingDefinitions), std::end(matchingDefinitions), std::back_inserter(options),
+                    [](const auto* current) { return current->name(); });
```

Comment:
More elegant could be:
```
const auto options = VectorUtils::map(matchingDefinitions,
                    [](const auto* current) { return current->name(); });
```
(Not that it is likely to matter here, but VectorUtils::map also allocates the correct sized vector before filling it, which I don't think std::transform/std::back_inserter can do?)

Reply:
Oh, that's fantastic. I'm going to use this everywhere. One question about the `map` function, is there any particular reason that it moves the return value instead of relying on RVO to construct it directly on the caller's stack?

Reply:
I think that was just a mistake on my part - putting the `std::move` there actually disables RVO, right?

Reply:
Yes. I'll change it in my other branch.